Desktop e-book: Chapter 5 (#17303)
dotnet/docs: docs/architecture/modernize-desktop/example-migration-to-net-core.md

of your existing application from .NET Framework to .NET Core.

We present a well-structured process you can follow and the most important
things to consider on each step.

We then document a step-by-step migration process for a sample desktop
application, both from WinForms and WPF versions.

## Migration Process Overview

The migration process consists of four sequential steps:

1. **Preparation**: Understand the dependencies the project has to have
   an idea of what is ahead. In this step, you take the current project into a
   state that simplifies the startup point for the migration.

2. **Migrate Project File:** .NET Core projects use the new SDK style projects.
  Create a new project file with this format or update the one you
   have to use the SDK style.

3. **Fix Code and Build:** Get the code building in .NET Core addressing API
   -level differences between .NET Framework and .NET Core. Update third-party
   packages to the ones that support .NET  Core if needed. WCF clients.

4. **Run and Test:** Forget about once it builds you are finished. There are
   differences that don't show up until run time, so you need to be sure
   everything works as expected.

> 

Author: I missed that when I was reading, the writing is so bad :) I'l rewrite the paragraph.

move your external packages references to the new format as mentioned above.

#### Update package references

Proceed to update the version of the packages you have found to be compatible as
depicted in previous section.

### Fix Code and Build

#### Microsoft.Windows.Compatibility

In case your applications depends on APIs that aren't available on .NET Core like
Registry, ACLs, WCF you have to include a reference to the
Microsoft.Windows.Compatibility package to add these Windows-specific APIs. They
work on .NET Core but aren't included as they aren't cross-platform.

There is a tool called API Analyzer
(<https://docs.microsoft.com/en-us/dotnet/standard/analyzers/api-analyzer>) that
help you identify not compatible APIs as long as you develop your code.

#### Use \#if defines

If you need different execution paths when targeting .NET Framework and .NET
Core, you should use compilation constants and code some \#if statements to keep
the same code base for both targets.

#### Technologies not available on .NET Core

Some technologies aren't available on .NET Core like:

* AppDomains
* Remoting
* Code Access Security
* WCF Server
* Windows Workflow

> Therefore, it is time to look for a replacement for those.

Reviewer: ```suggestion
Therefore, you'll need to look for a replacement for those if you're using them in your application.
```

> 

Reviewer: Should you link to https://docs.microsoft.com/en-us/dotnet/core/porting/net-framework-tech-unavailable for more info?

Author: Good idea!

#### Regenerate autogenerated clients

> If your application uses some autogenerated code like, for example, a WCF
> client, you may need to regenerate this code to target .NET Core. Sometimes you
> can find some missing references since it may not be included as default as part
> of the basic .Net Core assemblies. Using a tool like <https://apisof.net/> you
> can easily locate the assembly the missing reference lives in and add it from
> NuGet.

Reviewer: ```suggestion
If your application uses autogenerated code, such as a WCF
client, you may need to regenerate this code to target .NET Core. Sometimes, you
can find some missing references since it may not be included as default as part
of the basic .NET Core assemblies. Using a tool like <https://apisof.net/>, you
can easily locate the assembly the missing reference lives in and add it from
NuGet.
```

.NET Core, they should work well unless they call some unsupported APIs.

### Run and Test

Once you have your application building with no errors, you can start the last
step of the migration by testing every functionality.

In this final step, you can find a variety of issues depending on the complexity
of your application and the dependencies and APIs you are using.

For example, in the case you have some usage of configuration files (app.config)
you may find some errors at run time like Configuration Sections not present.
Using Microsoft.Extensions.Configuration should do the fix.

Another reason for errors is the use of BeginInvoke and EndInvoke because they
are supported on .NET Core. The reason they aren't supported on .NET Core is
they have a dependency on Remoting, which does not exist on .NET Core. To solve
this issue, try to use `await` when available or `Task.Run`.

You can use compatibility analyzers to let you identify APIs and code patterns
in your code that can potentially cause problems at run-time with .NET Core. You
can go to <http://github.com/dotnet/platform-compat> and use the Roselyn
analyzers.

> 

Author: We should call it .NET API Analyzer since we productized it with this name. Compatibility Analizer is the old internal name. but when we went public, it is refereed as .NET API Analyzer.

the project is targeting the .NET Core.

If we compile the project at this point, we find some errors related to the WCF
client reference. Since this is autogenerated code, we must regenerate it to
target .NET Core.

![Errors](media/example-migration-to-net-core/errors.png)

We can delete the `Reference.cs` file and generate a new Service Client.

Right-click over *Connected Services* and select the "*Add Connected Service*"
option.

![Add Connected Service](media/example-migration-to-net-core/add-connected-service.png)

This will open the Connected Services window where we select the Microsoft WCF
Web Service option.

![Connected Services window](media/example-migration-to-net-core/connected-services-window.png)

If we have the WCF Service in the same solution as we have, we can choose to
select the Discover option instead of specifying a service URL.

![Configure WCF Web Service Reference window](media/example-migration-to-net-core/configure-wcf-reference.png)

Once the service is located, the tool reflects the API contract implemented by
the service. We change the name of the Namespace to be eShopServiceReference:

> ![API contract](media/example-migration-to-net-core/api-contract.png)

Reviewer: The image name is a bit too short. Here's my suggestion:
```suggestion
![Screenshot the API contract and the namespace changed](media/example-migration-to-net-core/api-contract-namespace.png)
```

> 

Reviewer: The previous image would probably benefit of having the selection highlighted (others on the book would probably too)
Similar to this: https://docs.microsoft.com/en-us/dotnet/standard/analyzers/portability-analyzer#analyze-portability
Guidelines: https://review.docs.microsoft.com/en-us/help/contribute/contribute-how-to-create-screenshot?branch=master#add-red-boxes

> We click on *Finish* and after a while, we will see the generated code.

Reviewer: ```suggestion
Select the **Finish** button. After a while, you'll see the generated code.
```

> We can see three generated documents:

Reviewer: ```suggestion
You should see three autogenerated files:
```

> 1. Getting Started: just a link to GitHub to provide some info on WCF

Reviewer: Is this a link to a repo or to a file?

```suggestion
1. Getting Started: a link to GitHub to provide some information on WCF.
```

> 2. `ConnectedService.json`: configuration parameters to connect to the service.

Reviewer: ```suggestion
2. *ConnectedService.json*: configuration parameters to connect to the service.
```

> 3. `Reference.cs`: this is the actual WCF client code.

Reviewer: ```suggestion
3. *Reference.cs*: the actual WCF client code.
```

> ![Generated documents](media/example-migration-to-net-core/generated-documents.png)

Reviewer: ```suggestion
![Screenshot of the Solution Explorer window with the three autogenerated files](./media/example-migration-to-net-core/autogenerated-files.png)
```

> If we compile again, we see many errors coming from `.cs` files inside the Helper
> folder. This folder was present in the .NET Framework version but not included
> in the old .csproj. But with the new SDK project style every code file present
> underneath the project file location is included by default, meaning that the
> new .NET Core project tries to compile the files inside the Helper folder. Since
> it is not needed, we can safely delete it.

Reviewer: Here I'm assuming that it in the last sentence refers to the folder. Please review.

```suggestion
If you compile again, you'll see many errors coming from *.cs* files inside the *Helper*
folder. This folder was present in the .NET Framework version but not included
in the old .csproj. But with the new SDK-style project, every code file present
underneath the project file location is included by default. That is, the
new .NET Core project tries to compile the files inside the *Helper* folder. Since
that folder isn't needed, you can safely delete it.
```

> We compile again but when we execute the application, we see no images of
> products. The problem is that now the path to the files has changed slightly and
> we need to add another level of depth in the path, updating from:

Reviewer: ```suggestion
If you compile the project again and execute it, you won't see the product images.
The problem is that now the path to the files has slightly changed.
To fix this issue, you need to add another level of depth in the path, updating from:
```

> ```C#

Reviewer: ```suggestion
```csharp
```

string image_name = Environment.CurrentDirectory + "\\..\\..\\Assets\\Images\\Catalog\\" + catalogItems.Picturefilename;
```

> to

Reviewer: Not clear where this change should be made. It would be good to say where this line of code exists.

> ```C#

Reviewer: ```suggestion
```csharp
```

string image_name = Environment.CurrentDirectory + "\\..\\..\\..\\Assets\\Images\\Catalog\\" + catalogItems.Picturefilename;
```

> 

Reviewer: Not sure where this code exists but image_name and Picturefilename are not following coding guidelines in terms of naming conventions (https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/coding-style.md).

Author: it is https://github.com/dotnet-architecture/eShopOnContainers

> After this change, we check that the application launches and performs as
> expected on .NET Core.

Reviewer: ```suggestion
After this change, you can check that the application launches and runs as
expected on .NET Core.
```

## Migrating a WPF Application

> We will use the Shop.ClassicWPF sample application to perform the migration.
> Here is a screenshot of the app before migration.

Reviewer: ```suggestion
We'll use the `Shop.ClassicWPF` sample application to perform the migration.
The following image shows a screenshot of the app before migration:
```

> ![Sample app before migration](media/example-migration-to-net-core/before-migration.png)

Reviewer: suggesting a new file name

```suggestion
![Sample app before the migration](./media/example-migration-to-net-core/app-before-migration.png)
```

This application uses a local SQLExpress database to hold the product catalog
information. This database is accessed directly from the WPF application.

> We must first update the .csproj file to the new SDK style used by the .NET Core
> applications. We'll follow the same steps described in the Windows Forms
> migration, we unload the project, open the .csproj file and update its contents
> and finally reload the project.

Reviewer: ```suggestion
First, you must update the *.csproj* file to the new SDK style used by .NET Core
applications. You'll follow the same steps described in the Windows Forms
migration: you'll unload the project, open the *.csproj* file, update its contents,
and reload the project.
```

> In this case, we delete all the content of the .csproj file and replace it with:

Reviewer: ```suggestion
In this case, delete all the content of the *.csproj* file and replace it with the following code:
```

```xml
<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
    <PropertyGroup>
        <OutputType>WinExe</OutputType>
        <TargetFramework>netcoreapp3.0</TargetFramework>
        <UseWPF>true</UseWPF>
        <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
    </PropertyGroup>
</Project>
```

> If we reload the project and compile, we get the following error:

Reviewer: ```suggestion
If you reload the project and compile it, you'll get the following error:
```

> ![Compile error](media/example-migration-to-net-core/compile-error.png)

Reviewer: ```suggestion
![Screenshot of compilation errors on Visual Studio](media/example-migration-to-net-core/wpf-compilation-error.png)
```

> Since we have deleted all the .csproj contents we have lost a project reference
> specification present in the old project. We just need to add this line to
> .csproj to include it:

Reviewer: ```suggestion
Since you've deleted all the *.csproj* contents, you've lost a project reference
specification present in the old project. You just need to add this line to the
*.csproj* file to include the project reference back:
```

```xml
<ItemGroup>
    <ProjectReference Include="..\\eShop.SqlProvider\\eShop.SqlProvider.csproj" />
<ItemGroup>
```

> Alternatively, we can let Visual Studio help us by clicking on Add Reference
> option and select the project from the solution:

Reviewer: I'm a bit lost on this step. Wouldn't this be?

```suggestion
You can also let Visual Studio help you by right-clicking on the **Dependencies** node and
selecting **Add Project Reference**. Select the project from the solution and click **OK**:
```

> ![Reference Manager](media/example-migration-to-net-core/reference-manager.png)

Reviewer: ```suggestion
![Screenshot of the Reference Manager dialog with the eShop.SqlProvider project selected](./media/example-migration-to-net-core/reference-manager.png)
```

> Once we do it, the application compiles and executes as expected on .NET Core.

Reviewer: ```suggestion
Once you add the missing project reference, the application compiles and runs as expected on .NET Core.
```
